File: dp_revision/lecture_2/p6.java
public class p6
{
    public static int lds(int[] nums)
    {
        int[] maxVal=new int[nums.length];
        int ans=0;
        for(int val:nums)
        {
            int l=0;
            int h=ans;
            while(l<h)
            {
                int mid=(l+h)/2;
                if(maxVal[mid] > val)
                    l=mid+1;
                else
                    h=mid; 
            }
            maxVal[l]=val;
            if(l==ans)
                ans++;
        }
        return ans;
    }
    
    public static void main(String args[])
    {
        int[] nums={1,6,2,4,3};
        System.out.println(lds(nums));
    }
}

File: lecture_47/p6.cpp
int Solution::paint(int A, int B, vector<int> &C) {
    long long int min_time=0;
    long long int max_time=0;
    for(long long int val:C)
    {
        min_time=max(min_time,val);
        max_time+=val;
    }
    long long int l=0;
    long long int h=max_time;
    long long int ans=0;
    while(l<=h)
    {
        long long int mid=(l+h)/2;
        long long int temp=0;
        int pr=1;
        for(int val:C)
        {
            if(temp+val > mid)
            {
                pr++;
                temp=0;
            }
            temp+=val;
        }
        if(pr <= A)
        {
            ans=mid;
            h=mid-1;
        }
        else
        {
            l=mid+1;
        }
    }
    return (ans*B)%10000003;
}
